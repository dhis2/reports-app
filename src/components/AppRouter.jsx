import React from 'react'
import { Route, Switch } from 'react-router-dom'
import {
    sections,
    DATA_SET_REPORT_SECTION_KEY,
    ORG_UNIT_DIST_REPORT_SECTION_KEY,
    REPORTING_RATE_SUMMARY_SECTION_KEY,
    RESOURCE_SECTION_KEY,
    STANDARD_REPORT_SECTION_KEY,
} from '../config/sections.config.js'
import { DataSetReport } from '../pages/DataSetReport.jsx'
import Home from '../pages/home/Home.jsx'
import { OrganisationUnitDistributionReport } from '../pages/OrganisationUnitDistributionReport.jsx'
import { ReportingRateSummary } from '../pages/ReportingRateSummary.jsx'
import { AddEditResource } from '../pages/resource/AddEditResource.jsx'
import { Resource } from '../pages/Resource.jsx'
import { ConnectedAddEditStdReport } from '../pages/standard-report/AddEditStdReport.jsx'
import StyledHtmlReport from '../pages/standard-report/StyledHtmlReport.jsx'
import { StandardReport } from '../pages/StandardReport.jsx'
import NoMatch from './NoMatch.jsx'

const standardReportPath = sections[STANDARD_REPORT_SECTION_KEY].path
const resourcePath = sections[RESOURCE_SECTION_KEY].path
// Used to ensure we are matching the pattern of an actual id
const ID_REGEXP = '[a-zA-Z][a-zA-Z0-9]{10}'

const AppRouter = () => (
    <main>
        <Switch>
            <Route key="home" exact path="/" component={Home} />
            <Route
                exact
                key={STANDARD_REPORT_SECTION_KEY}
                path={standardReportPath}
                component={StandardReport}
            />
            <Route
                exact
                key={`${STANDARD_REPORT_SECTION_KEY}-viewHTMLReport`}
                path={`${standardReportPath}/view/:id(${ID_REGEXP})`}
                component={StyledHtmlReport}
            />
            <Route
                exact
                key={`${STANDARD_REPORT_SECTION_KEY}-addEdit`}
                // /edit/id or /new
                path={`${standardReportPath}/:mode/:id(${ID_REGEXP})?`}
                component={ConnectedAddEditStdReport}
            />
            <Route
                exact
                key={DATA_SET_REPORT_SECTION_KEY}
                path={sections[DATA_SET_REPORT_SECTION_KEY].path}
                component={DataSetReport}
            />
            <Route
                exact
                key={REPORTING_RATE_SUMMARY_SECTION_KEY}
                path={sections[REPORTING_RATE_SUMMARY_SECTION_KEY].path}
                component={ReportingRateSummary}
            />
            <Route
                exact
                key={RESOURCE_SECTION_KEY}
                path={resourcePath}
                component={Resource}
            />
            <Route
                exact
                key={`${RESOURCE_SECTION_KEY}-addEdit`}
                // /edit/id or /new
                path={`${resourcePath}/:mode/:id(${ID_REGEXP})?`}
                component={AddEditResource}
            />
            <Route
                exact
                key={ORG_UNIT_DIST_REPORT_SECTION_KEY}
                path={sections[ORG_UNIT_DIST_REPORT_SECTION_KEY].path}
                component={OrganisationUnitDistributionReport}
            />
            <Route key="no-match-route" component={NoMatch} />
        </Switch>
    </main>
)

export default AppRouter
