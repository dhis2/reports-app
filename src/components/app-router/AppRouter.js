/* React */
import React from 'react'

/* React Router */
import { Route, Switch } from 'react-router-dom'

/* App components */
import Home from '../../pages/home/Home'
import NoMatch from './NoMatch'

/* App context */
import AppContext from '../../pages/AppContext'

/* App configs */
import { sections } from '../../pages/sections.conf'

const AppRouter = () => {
    const routes = sections.map(section => {
        const routeRender = () => {
            const Page = section.component
            return (
                <AppContext.Consumer>
                    {appContext => (
                        <Page
                            d2={appContext.d2}
                            sectionKey={section.key}
                            {...appContext.pageState}
                        />
                    )}
                </AppContext.Consumer>
            )
        }
        return (
            <Route
                key={section.key}
                exact
                path={section.path}
                render={routeRender}
            />
        )
    })

    /* Home route */
    const homeRouteRender = () => <Home sectionKey="home" />

    routes.push(<Route key="home" exact path="/" render={homeRouteRender} />)

    /* No Match Route */
    routes.push(<Route key="no-match-route" component={NoMatch} />)

    return (
        <main>
            <Switch>{routes}</Switch>
        </main>
    )
}

export default AppRouter
