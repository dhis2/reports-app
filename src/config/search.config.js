export const DEBOUNCE_DELAY = 500
