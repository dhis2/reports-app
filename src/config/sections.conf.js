/* Page Components */
import { ConnectedStandardReport } from '../pages/StandardReport'
import { ConnectedDataSetReport } from '../pages/DataSetReport'
import { ConnectedReportingRateSummary } from '../pages/ReportingRateSummary'
import { ConnectedResource } from '../pages/resource/Resource'
import { ConnectedOrganisationUnitDistributionReport } from '../pages/organisation-unit-distribution-report/OrganisationUnitDistributionReport'
import { ConnectedDataApproval } from '../pages/data-approval/DataApproval'

/* i18n */
import { i18nKeys } from '../utils/i18n/i18nKeys'

export const STANDARD_REPORT_SECTION_KEY = 'standard-report'
export const DATA_SET_REPORT_SECTION_KEY = 'data-set-report'
export const REPORTING_RATE_SUMMARY_SECTION_KEY = 'reporting-rate-summary'
export const RESOURCE_SECTION_KEY = 'resource'
export const ORGANISATION_UNIT_DISTRIBUTION_REPORT_SECTION_KEY =
    'organisation-unit-distribution-report'
export const DATA_APPROVAL_SECTION_KEY = 'data-approval'

export const DEBOUNCE_DELAY = 500

export const sections = [
    {
        key: STANDARD_REPORT_SECTION_KEY,
        path: '/standard-report',
        component: ConnectedStandardReport,
        info: {
            label: i18nKeys.standardReport.homeLabel,
            icon: 'bar_chart',
            description: i18nKeys.standardReport.description,
            actionText: i18nKeys.standardReport.homeAction,
            docs: 'using_reporting_standard_reports',
        },
    },
    {
        key: DATA_SET_REPORT_SECTION_KEY,
        path: '/data-set-report',
        component: ConnectedDataSetReport,
        info: {
            label: i18nKeys.dataSetReport.homeLabel,
            icon: 'assignment',
            description: i18nKeys.dataSetReport.description,
            actionText: i18nKeys.dataSetReport.homeAction,
            docs: 'using_reporting_dataset_reports',
        },
    },
    {
        key: REPORTING_RATE_SUMMARY_SECTION_KEY,
        path: '/reporting-rate-summary',
        component: ConnectedReportingRateSummary,
        info: {
            label: i18nKeys.reportingRateSummary.homeLabel,
            icon: 'format_list_bulleted',
            description: i18nKeys.reportingRateSummary.description,
            actionText: i18nKeys.reportingRateSummary.homeAction,
            docs: 'using_reporting_reporting_rate_summary',
        },
    },
    {
        key: RESOURCE_SECTION_KEY,
        path: '/resource',
        component: ConnectedResource,
        info: {
            label: i18nKeys.resource.homeLabel,
            icon: 'cloud_upload',
            description: i18nKeys.resource.description,
            actionText: i18nKeys.resource.homeAction,
            docs: 'using_reporting_resources',
        },
    },
    {
        key: ORGANISATION_UNIT_DISTRIBUTION_REPORT_SECTION_KEY,
        path: '/organisation-unit-distribution-report',
        component: ConnectedOrganisationUnitDistributionReport,
        info: {
            label: i18nKeys.organisationUnitDistributionReport.homeLabel,
            icon: 'device_hub',
            description:
                i18nKeys.organisationUnitDistributionReport.description,
            actionText: i18nKeys.organisationUnitDistributionReport.homeAction,
            docs: 'using_reporting_orgunit_distribution_reports',
        },
    },
    {
        key: DATA_APPROVAL_SECTION_KEY,
        path: '/data-approval',
        component: ConnectedDataApproval,
        info: {
            label: i18nKeys.dataApproval.homeLabel,
            icon: 'playlist_add_check',
            description: i18nKeys.dataApproval.description,
            actionText: i18nKeys.dataApproval.homeAction,
            docs: 'data_approval',
        },
    },
]

export const getDocsKeyForSection = sectionKey => {
    for (let i = 0; i < sections.length; i++) {
        const section = sections[i]
        if (section.key === sectionKey) {
            return section.info.docs
        }
    }

    return ''
}
