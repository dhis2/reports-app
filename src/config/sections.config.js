import i18n from '../locales'

export const STANDARD_REPORT_SECTION_KEY = 'standard-report'
export const DATA_SET_REPORT_SECTION_KEY = 'data-set-report'
export const REPORTING_RATE_SUMMARY_SECTION_KEY = 'reporting-rate-summary'
export const RESOURCE_SECTION_KEY = 'resource'
export const ORG_UNIT_DIST_REPORT_SECTION_KEY =
    'organisation-unit-distribution-report'
export const DATA_APPROVAL_SECTION_KEY = 'data-approval'

export const sectionOrder = [
    STANDARD_REPORT_SECTION_KEY,
    DATA_SET_REPORT_SECTION_KEY,
    REPORTING_RATE_SUMMARY_SECTION_KEY,
    RESOURCE_SECTION_KEY,
    ORG_UNIT_DIST_REPORT_SECTION_KEY,
]

export const sections = {
    [STANDARD_REPORT_SECTION_KEY]: {
        key: STANDARD_REPORT_SECTION_KEY,
        path: `/${STANDARD_REPORT_SECTION_KEY}`,
        info: {
            label: i18n.t('Standard Report'),
            icon: 'bar_chart',
            description: i18n.t(
                'View and add reports based on the JasperReports library. These can be based on report tables and can be designed in iReport.'
            ),
            actionText: i18n.t('View Reports'),
            docs: 'using_reporting_standard_reports',
        },
    },
    [DATA_SET_REPORT_SECTION_KEY]: {
        key: DATA_SET_REPORT_SECTION_KEY,
        path: `/${DATA_SET_REPORT_SECTION_KEY}`,
        info: {
            label: i18n.t('Data Set Report'),
            icon: 'assignment',
            description: i18n.t(
                'View data set reports. These reports are based on data entry screens and will produce a report with aggregated data.'
            ),
            actionText: i18n.t('Get Report'),
            docs: 'using_reporting_dataset_reports',
        },
    },
    [REPORTING_RATE_SUMMARY_SECTION_KEY]: {
        key: REPORTING_RATE_SUMMARY_SECTION_KEY,
        path: `/${REPORTING_RATE_SUMMARY_SECTION_KEY}`,
        info: {
            label: i18n.t('Reporting Rate Summary'),
            description: i18n.t(
                'Browse the reporting rates of data sets by organisation unit and period based on various criteria for submission.'
            ),
            actionText: i18n.t('Get Report'),
            icon: 'format_list_bulleted',
            docs: 'using_reporting_reporting_rate_summary',
        },
    },
    [RESOURCE_SECTION_KEY]: {
        key: RESOURCE_SECTION_KEY,
        path: `/${RESOURCE_SECTION_KEY}`,
        info: {
            label: i18n.t('Resource'),
            description: i18n.t(
                'View and add resources. These resources can be uploaded documents or URLs on the web.'
            ),
            actionText: i18n.t('View Resources'),
            icon: 'cloud_upload',
            docs: 'using_reporting_resources',
        },
    },
    [ORG_UNIT_DIST_REPORT_SECTION_KEY]: {
        key: ORG_UNIT_DIST_REPORT_SECTION_KEY,
        path: `/${ORG_UNIT_DIST_REPORT_SECTION_KEY}`,
        info: {
            label: i18n.t('Org unit distribution report'),
            description: i18n.t(
                'Browse the organisation unit distribution report based on the organisation unit group sets and its groups.'
            ),
            actionText: i18n.t('Get Report'),
            icon: 'device_hub',
            docs: 'using_reporting_orgunit_distribution_reports',
        },
    },
}
