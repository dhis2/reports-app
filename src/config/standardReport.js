export const reportParameterOptions = [
    {
        value: 'paramReportingPeriod',
        label: 'Reporting Period',
    },
    {
        value: 'paramOrganisationUnit',
        label: 'Organisation Unit',
    },
]
