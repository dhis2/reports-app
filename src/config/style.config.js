export const FONT_FAMILY = "'Roboto', sans-serif"
export const BASE_COLOR = '39, 102, 150' // DHIS2 blue
