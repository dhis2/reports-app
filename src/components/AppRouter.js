import { Route, Switch } from 'react-router-dom'
import React from 'react'
// import PropTypes from 'prop-types'

import {
    DATA_SET_REPORT_SECTION_KEY,
    ORG_UNIT_DIST_REPORT_SECTION_KEY,
    REPORTING_RATE_SUMMARY_SECTION_KEY,
    RESOURCE_SECTION_KEY,
    STANDARD_REPORT_SECTION_KEY,
} from '../config/sections.config'
import { DataSetReport } from '../pages/DataSetReport'
import { OrganisationUnitDistributionReport } from '../pages/OrganisationUnitDistributionReport'
import { ReportingRateSummary } from '../pages/ReportingRateSummary'
import { Resource } from '../pages/Resource'
import { StandardReport } from '../pages/StandardReport'
import StyledHtmlReport from '../pages/standard-report/StyledHtmlReport'
import { sections } from '../config/sections.config'
import Home from '../pages/home/Home'
import NoMatch from './NoMatch'
import { ConnectedAddEditStdReport } from '../pages/standard-report/AddEditStdReport'
import { AddEditResource } from '../pages/resource/AddEditResource'

const standardReportPath = sections[STANDARD_REPORT_SECTION_KEY].path
const resourcePath = sections[RESOURCE_SECTION_KEY].path

const AppRouter = () => (
    <main>
        <Switch>
            <Route key="home" exact path="/" component={Home} />
            <Route
                exact
                key={STANDARD_REPORT_SECTION_KEY}
                path={standardReportPath}
                component={StandardReport}
            />
            <Route
                exact
                key={`${STANDARD_REPORT_SECTION_KEY}-viewHTMLReport`}
                path={`${standardReportPath}/view/:id`}
                component={StyledHtmlReport}
            />
            <Route
                exact
                key={`${STANDARD_REPORT_SECTION_KEY}-addEdit`}
                // /edit/id or /new
                path={`${standardReportPath}/:mode/:id?`}
                component={ConnectedAddEditStdReport}
            />
            <Route
                exact
                key={DATA_SET_REPORT_SECTION_KEY}
                path={sections[DATA_SET_REPORT_SECTION_KEY].path}
                component={DataSetReport}
            />
            <Route
                exact
                key={REPORTING_RATE_SUMMARY_SECTION_KEY}
                path={sections[REPORTING_RATE_SUMMARY_SECTION_KEY].path}
                component={ReportingRateSummary}
            />
            <Route
                exact
                key={RESOURCE_SECTION_KEY}
                path={resourcePath}
                component={Resource}
            />
            <Route
                exact
                key={`${RESOURCE_SECTION_KEY}-addEdit`}
                // /edit/id or /new
                path={`${resourcePath}/:mode/:id?`}
                component={AddEditResource}
            />
            <Route
                exact
                key={ORG_UNIT_DIST_REPORT_SECTION_KEY}
                path={sections[ORG_UNIT_DIST_REPORT_SECTION_KEY].path}
                component={OrganisationUnitDistributionReport}
            />
            <Route key="no-match-route" component={NoMatch} />
        </Switch>
    </main>
)

export default AppRouter
