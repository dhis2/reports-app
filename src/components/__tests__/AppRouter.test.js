import { Route, Switch } from 'react-router-dom'
import React from 'react'

import { shallow } from 'enzyme'

import AppRouter from '../AppRouter'
import { sectionOrder } from '../../config/sections.config'

jest.mock('@dhis2/d2-ui-org-unit-tree', () => ({ OrgUnitTree: 'OrgUnitTree' }))

jest.mock('@dhis2/d2-ui-org-unit-tree', () => 'OrgUnitTree')

const ownShallow = () => shallow(<AppRouter />)

it('AppRouter renders without crashing', () => {
    ownShallow()
})

it('AppRouter renders the correct number of Route', () => {
    const wrapper = ownShallow()

    // Pages plus home and no match route
    expect(wrapper.find(Route)).toHaveLength(sectionOrder.length + 2)
})
