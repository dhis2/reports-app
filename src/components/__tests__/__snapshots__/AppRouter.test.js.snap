// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Matches the snapshot 1`] = `
<main>
  <Switch>
    <Route
      component={[Function]}
      exact={true}
      key="home"
      path="/"
    />
    <Route
      component={
        Object {
          "$$typeof": Symbol(react.memo),
          "WrappedComponent": [Function],
          "compare": null,
          "displayName": "Connect(StandardReport)",
          "type": [Function],
        }
      }
      exact={true}
      key="standard-report"
      path="/standard-report"
    />
    <Route
      component={
        Object {
          "$$typeof": Symbol(react.memo),
          "WrappedComponent": [Function],
          "compare": null,
          "displayName": "Connect(StyledHtmlReport)",
          "type": [Function],
        }
      }
      exact={true}
      key="standard-report-viewHTMLReport"
      path="/standard-report/view/:id"
    />
    <Route
      component={
        Object {
          "$$typeof": Symbol(react.memo),
          "WrappedComponent": [Function],
          "compare": null,
          "displayName": "Connect(AddEditStandardReport)",
          "type": [Function],
        }
      }
      exact={true}
      key="standard-report-addEdit"
      path="/standard-report/:mode/:id?"
    />
    <Route
      component={
        Object {
          "$$typeof": Symbol(react.memo),
          "WrappedComponent": [Function],
          "compare": null,
          "displayName": "Connect(DataSetReport)",
          "type": [Function],
        }
      }
      exact={true}
      key="data-set-report"
      path="/data-set-report"
    />
    <Route
      component={
        Object {
          "$$typeof": Symbol(react.memo),
          "WrappedComponent": [Function],
          "compare": null,
          "displayName": "Connect(ReportingRateSummary)",
          "type": [Function],
        }
      }
      exact={true}
      key="reporting-rate-summary"
      path="/reporting-rate-summary"
    />
    <Route
      component={
        Object {
          "$$typeof": Symbol(react.memo),
          "WrappedComponent": [Function],
          "compare": null,
          "displayName": "Connect(Resource)",
          "type": [Function],
        }
      }
      exact={true}
      key="resource"
      path="/resource"
    />
    <Route
      component={
        Object {
          "$$typeof": Symbol(react.memo),
          "WrappedComponent": [Function],
          "compare": null,
          "displayName": "Connect(AddEditResource)",
          "type": [Function],
        }
      }
      exact={true}
      key="resource-addEdit"
      path="/resource/:mode/:id?"
    />
    <Route
      component={
        Object {
          "$$typeof": Symbol(react.memo),
          "WrappedComponent": [Function],
          "compare": null,
          "displayName": "Connect(OrganisationUnitDistributionReport)",
          "type": [Function],
        }
      }
      exact={true}
      key="organisation-unit-distribution-report"
      path="/organisation-unit-distribution-report"
    />
    <Route
      component={[Function]}
      key="no-match-route"
    />
  </Switch>
</main>
`;
